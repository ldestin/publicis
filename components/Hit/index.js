import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import dayjs from '../../utils/dayjs'

export const getHostName = url => {
  try {
    return new URL(url).hostname
  } catch {
    return undefined
  }
}

const Hit = ({ hit, color, onHideClick: customOnHideClick, onUpvoteClick: customOnUpvoteClick }) => {
  const { num_comments, points, title, url, author, created_at } = hit
  const hostName = getHostName(url)
  const onHideClick = useCallback(() => {
    if (customOnHideClick) customOnHideClick(hit)
  }, [hit, customOnHideClick])
  const onUpvoteClick = useCallback(() => {
    if (customOnUpvoteClick) customOnUpvoteClick(hit)
  }, [hit, customOnHideClick])

  return (
    <div className="hit">
      <span className="line numbers">
        <span className="col nb-comments">{num_comments || 0}</span>{' '}
        <span className="label">comment{num_comments > 1 ? 's' : ''}</span>
        <span className="col points" style={{ color }}>{points || 0}</span>{' '}
        <span className="label">point{points > 1 ? 's' : ''}</span>
      </span>
      <div className="line">
        <div className={classnames('col button upvote-button', { hidden: hit.blockVote })} onClick={onUpvoteClick} />{' '}
        <span className="col description">
          {title && <><a href={url}>{title}</a>{' '}</>}
          {hostName && <span className="small grey">({hostName}){' '}</span>}
          {author && <><span className="small grey">by </span><span>{author}</span>{' '}</>}
          {created_at && <><span className="small grey">{dayjs(created_at).fromNow()}</span>{' '}</>}
          <span className="button hide-button small" onClick={onHideClick}><span className="grey">[</span>hide<span className="grey">]</span></span>
        </span>
      </div>
    </div>
  )
}

Hit.defaultProps = {
  onHideClick: undefined,
}

Hit.propTypes = {
  hit: PropTypes.shape({
    num_comments: PropTypes.number,
    points: PropTypes.number,
    title: PropTypes.string,
    url: PropTypes.string,
    author: PropTypes.string,
    created_at: PropTypes.string,
  }).isRequired,
  color: PropTypes.string.isRequired,
  onHideClick: PropTypes.func,
}

export default Hit
