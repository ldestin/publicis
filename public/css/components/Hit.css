.hit {
  display: flex;
  padding: 4px 2rem;
}

.hit:nth-child(even) {
  background: rgba(0, 0, 0, .1);
}

.hit .col:not(:first-child) {
  margin-left: 1rem;
}

.hit .line {
  display: contents;
}

.hit .line:not(:last-child) {
  margin-top: 1rem;
}

.hit .numbers {
  justify-content: center;
}

.hit .label {
  display: none;
  margin-left: 1rem;
}

.hit .nb-comments,
.hit .points {
  flex-shrink: 0;
  width: 3rem;
  text-align: end;
}

.hit .upvote-button {
  margin-left: .5rem;
}

.hit .upvote-button::before {
  content: 'â–²';
  color: var(--grey);
}

@media (max-width: 700px) {
  .hit {
    flex-direction: column-reverse;
  }

  .hit .label {
    display: inline;
  }

  .hit .line {
    display: flex;
    align-items: center;
  }

  .hit .numbers {
    font-size: var(--small-font-size);
  }
}
