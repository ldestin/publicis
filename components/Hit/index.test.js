import React from 'react'
import { render } from '@testing-library/react';
import Hit, { getHostName } from '.';

const mockDayJs = {
  fromNow() {
    return '20 hours ago'
  }
}
jest.mock('../../utils/dayjs', () => () => mockDayJs)

describe('getHostName function', () => {
  it('should return hostName from the specified url', () => {
    const hostName = getHostName('http://www.hostname.com/path?param=value')

    expect(hostName).toBe('www.hostname.com')
  })

  it('should return undefined with invalid url', () => {
    const hostName = getHostName('www.hostname.com')

    expect(hostName).toBe(undefined)
  })
})

describe('Hit component', () => {
  const props = {
    hit: {
      num_comments: 6,
      points: 13,
      title: 'title',
      url: 'url',
      author: 'author',
      created_at: '2020-11-10T16:55:58.000Z',
    },
    color: '#f00',
    onHideClick: jest.fn(),
    onUpvoteClick: jest.fn(),
  }

  it('should render correctly', () => {
    const { container } = render(<Hit {...props} />)

    expect(container).toMatchSnapshot()
  })

  it('should render correctly with empty hit', () => {
    const { container } = render(<Hit {...props} hit={{}} />)

    expect(container).toMatchSnapshot()
  })

  it('should call onHideClick correctly', () => {
    const { container } = render(<Hit {...props} />)
    const hideButton = container.querySelector('.hide-button')

    hideButton.click()

    expect(props.onHideClick).toHaveBeenCalledWith(props.hit)
  })

  it('should call onUpvoteClick correctly', () => {
    const { container } = render(<Hit {...props} />)
    const hideButton = container.querySelector('.upvote-button')

    hideButton.click()

    expect(props.onUpvoteClick).toHaveBeenCalledWith(props.hit)
  })
})
