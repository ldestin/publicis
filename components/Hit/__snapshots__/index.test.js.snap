// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hit component should render correctly 1`] = `
<div>
  <div
    class="hit"
  >
    <span
      class="line numbers"
    >
      <span
        class="col nb-comments"
      >
        6
      </span>
       
      <span
        class="label"
      >
        comment
        s
      </span>
      <span
        class="col points"
        style="color: rgb(255, 0, 0);"
      >
        13
      </span>
       
      <span
        class="label"
      >
        point
        s
      </span>
    </span>
    <div
      class="line"
    >
      <div
        class="col button upvote-button"
      />
       
      <span
        class="col description"
      >
        <a
          href="url"
        >
          title
        </a>
         
        <span
          class="small grey"
        >
          by 
        </span>
        <span>
          author
        </span>
         
        <span
          class="small grey"
        >
          20 hours ago
        </span>
         
        <span
          class="button hide-button small"
        >
          <span
            class="grey"
          >
            [
          </span>
          hide
          <span
            class="grey"
          >
            ]
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;

exports[`Hit component should render correctly with empty hit 1`] = `
<div>
  <div
    class="hit"
  >
    <span
      class="line numbers"
    >
      <span
        class="col nb-comments"
      >
        0
      </span>
       
      <span
        class="label"
      >
        comment
        
      </span>
      <span
        class="col points"
        style="color: rgb(255, 0, 0);"
      >
        0
      </span>
       
      <span
        class="label"
      >
        point
        
      </span>
    </span>
    <div
      class="line"
    >
      <div
        class="col button upvote-button"
      />
       
      <span
        class="col description"
      >
        <span
          class="button hide-button small"
        >
          <span
            class="grey"
          >
            [
          </span>
          hide
          <span
            class="grey"
          >
            ]
          </span>
        </span>
      </span>
    </div>
  </div>
</div>
`;
